modules/apps/message-boards/message-boards-uad-test/src/testIntegration/java/com/liferay/message/boards/uad/display/test/MBThreadUADDisplayTest.java

```diff
@@ -0,0 +1,255 @@
+		new LiferayIntegrationTestRule();
+
+	@Test
+	public void testGetParentContainerId() throws Exception {
+		assertParentContainerId(MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID);
+
+		MBCategory mbCategory = _addCategory(
+			MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID);
+
+		assertParentContainerId(mbCategory.getCategoryId());
+	}
+
+	@Test
+	public void testGetTopLevelContainer() throws Exception {
+
+		// A thread that is an immediate child of the given parent should
+		// return itself
+
+		MBThread mbThread = _addThread(
+			MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID);
+
+		Assert.assertEquals(
+			mbThread,
+			_getTopLevelContainer(
+				MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID, mbThread));
+
+		MBCategory mbCategoryA = _addCategory(
+			MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID);
+
+		MBThread mbThreadA = _addThread(mbCategoryA.getCategoryId());
+
+		Assert.assertEquals(
+			mbThreadA,
+			_getTopLevelContainer(mbCategoryA.getCategoryId(), mbThreadA));
+
+		// A message whose thread is an immediate child of the given parent
+		// should return its own thread
+
+		MBMessage mbMessage = _addMessage(
+			MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID,
+			mbThread.getThreadId());
+
+		Assert.assertEquals(
+			mbThread,
+			_getTopLevelContainer(
+				MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID, mbMessage));
+
+		MBMessage mbMessageA = _addMessage(
+			mbCategoryA.getCategoryId(), mbThreadA.getThreadId());
+
+		Assert.assertEquals(
+			mbThreadA,
+			_getTopLevelContainer(mbCategoryA.getCategoryId(), mbMessageA));
+
+		// A thread that is the parent should return null
+
+		Assert.assertNull(
+			_uadDisplay.getTopLevelContainer(
+				MBThread.class, mbThread.getThreadId(), mbThread));
+
+		// A thread that is not a descendant of the given parent should return
+		// null
+
+		MBCategory mbCategoryB = _addCategory(
+			MBCategoryConstants.DEFAULT_PARENT_CATEGORY_ID);
+
+		MBThread mbThreadB = _addThread(mbCategoryB.getCategoryId());
+
+		Assert.assertNull(
+			_getTopLevelContainer(mbCategoryA.getCategoryId(), mbThreadB));
+
+		// A message that is not a descendant of the given parent should
+		// return null
+
+		MBMessage mbMessageB = _addMessage(
+			mbCategoryB.getCategoryId(), mbThreadB.getThreadId());
+
+		Assert.assertNull(
+			_getTopLevelContainer(mbCategoryA.getCategoryId(), mbMessageB));
+
+		// A thread that is a non-immediate descendant of the given parent
+		// should return null
+
+		MBCategory mbCategoryBA = _addCategory(mbCategoryB.getCategoryId());
+
+		MBThread mbThreadBA = _addThread(mbCategoryBA.getCategoryId());
+
+		Assert.assertNull(
+			_getTopLevelContainer(mbCategoryB.getCategoryId(), mbThreadBA));
+
+		// A message that is a non-immediate descendant of the given parent
+		// should return null
+
+		MBMessage mbMessageBA = _addMessage(
+			mbCategoryBA.getCategoryId(), mbThreadBA.getThreadId());
+
+		Assert.assertNull(
+			_getTopLevelContainer(mbCategoryB.getCategoryId(), mbMessageBA));
```

Comment:
I'm struggling to understand these last two conditions.  Why would a non-immediate descendant return null instead of some sort of ID?